Sprint_6

blinchik3000 opened the pull request:
основная задача + доп

blinchik3000 committed:
add plugins and dependencies to pom
add new Class LionAlex with its methods
add dependency injection to class Lion - new Interface
cover all classes by unit tests

blinchik3000 committed:
add jacoco report

pomegranate282922 commented on src/main/java/com/example/Feline.java:
```diff
@@ -2,22 +2,32 @@
-public class Feline extends Animal implements Predator {
+public class Feline extends Animal implements Felines, Predator {
+
+    private Animal animal;
```
В принципе, класс Feline можно никак не менять.
Сейчас ты и наследуешься от Animal и поле класса Animal у тебя в классе. Перебор.
Советую вернуть все в исходное состояние.

pomegranate282922 commented on src/main/java/com/example/Lion.java:
```diff
@@ -12,11 +12,15 @@ public Lion(String sex) throws Exception {
             hasMane = false;
         } else {
-            throw new Exception("Используйте допустимые значения пола животного - самей или самка");
+            throw new Exception("Используйте допустимые значения пола животного - самец или самка");
```
Этот конструктор в итоге создаёт объект с неинициализированным полем feline. Это нехорошо, так как может вызвать NPE. Исправь, пожалуйста

blinchik3000 replied:
немного не понял - поправил опечатку в изначальном коде в данной строке, данный конструктор не создавал - задания его трогать вроде тоже не было

pomegranate282922 commented on src/test/java/FelinesTest/CatTest.java:
```diff
@@ -0,0 +1,36 @@
+package FelinesTest;
```
Пакеты именуем строчными буквами без camelCase

pomegranate282922 commented on src/test/java/FelinesTest/CatTest.java:
```diff
@@ -0,0 +1,36 @@
+
+    @Test
+    public void checkCatGetSound() {
+        Cat cat = new Cat(new Feline());
```
А почему здесь не использовать мок?

pomegranate282922 commented on src/test/java/FelinesTest/CatTest.java:
```diff
@@ -0,0 +1,36 @@
+        Assert.assertEquals("Кошка говорит неверно", "Мяу", catSound);
+    }
+
+    @Mock
```
Сначала перечисляем все поля класса, потом все методы

pomegranate282922 commented on src/test/java/FelinesTest/LionHasManeTest.java:
```diff
@@ -0,0 +1,45 @@
+            }
+
+        } catch (Exception e){
+            Assert.assertEquals("Ошибка не соответствует ожидаемой","java.lang.Exception:"+" Используйте допустимые значения пола животного - самец или самка",e.toString());
```
Это плохо тем, что если это исключение выпадет на самце или самке, тест будет зелёным. А это баг.

pomegranate282922 replied:
Разнести позитивные и негативные кейсы по разным классам. Негативный кейс на исключения сделай с использованием аннотации и assertThrows ( опционально)

pomegranate282922 commented on src/test/java/FelinesTest/LionTest.java:
```diff
@@ -0,0 +1,37 @@
+        Assert.assertEquals("Количество котят не соответствует ожидаемому",1,lion.getKittens());
+    }
+    @Mock
+    Feline feline1;
```
Так, а зачем два мока? Чем первый не подходит?

pomegranate282922 reviewed (commented):
Хорошая работа!
оставила комментарии, что нужно поправить.
В классе Lion - конструкторы. Класс Feline верни в исходное состояние или переработай корректно наследование.
В тестах раздели негативные и позитивные кейсы в Lion. 
остальное по мелочи.

blinchik3000 committed:
fix
get back to first Feline.java realization
package name to lower
refactor Mocks
refactor exception test

blinchik3000 committed:
fix
initialize variable for Lion constructor

pomegranate282922 commented on src/main/java/com/example/Lion.java:
```diff
@@ -5,18 +5,23 @@
-            throw new Exception("Используйте допустимые значения пола животного - самей или самка");
+            throw new Exception("Используйте допустимые значения пола животного - самец или самка");
         }
+        feline=new Feline();
```
Не, так ты опять вернулся к неизолированности lion от feline. Нам нельзя создавать feline внутри lion, он должен придти туда готовым извне.

pomegranate282922 commented on src/test/java/felinestest/CatTest.java:
```diff
@@ -0,0 +1,37 @@
+    @Mock
+    Feline feline;
+    @Mock
+    Feline predator;
```
А зачем два поля? Почему не использовать одно?

pomegranate282922 commented on src/test/java/felinestest/FelineTest.java:
```diff
@@ -0,0 +1,43 @@
+    public void checkFelineEatMeat() throws Exception {
+        Feline feline = new Feline(animal);
+        Mockito.when(animal.getFood("Хищник")).thenReturn(List.of("Животные", "Птицы", "Рыба"));
+        Assert.assertEquals("У кошачьих другой рацион", List.of("Животные", "Птицы", "Рыба"),feline.eatMeat());
```
Дважды создаёшь один и тот же список. Лучше переиспользовать

pomegranate282922 reviewed (commented):
Ok! Осталась маленькая, но важная помарка в конструкторе Lion.

blinchik3000 committed:
fix:
fix dependency injection in Lion
delete unnecessary mock in CatTest
create variable for reused List

pomegranate282922 reviewed (approved):
Супер!
